Budget csh lyh

wholemann opened the pull request.

wholemann committed:
Initial commit

wholemann committed:
Solve find maximun supportable budget problem by Kotlin

wholeman
영한

ahastudio commented on budget-csh-lyh/src/test/kotlin/com/codingdojo/budget/SolutionTest.kt:
```diff
@@ -0,0 +1,24 @@
+
+tailrec fun solution(d: IntArray, budget: Int): Int = when {
+    d.fold(0, { total, next -> total + next }) <= budget -> d.size
+    else -> solution(d.dropLast(1).toIntArray(), budget)
```
매우 중요한 이유가 없다면 그냥 `Array` 대신 `List`를 쓰시는 게 변환 없이 깔끔하게 쓸 수 있습니다. `d`의 타입을 `Collection`으로 잡아보세요.

wholemann replied:
아 그렇게 하면 되겠네요! 감사합니다

ahastudio replied:
그런데 `Collection`엔 `dropLast`가 없네요...

wholemann committed:
Add missing array sorting

ahastudio commented on budget-csh-lyh/src/test/kotlin/com/codingdojo/budget/SolutionTest.kt:
```diff
@@ -0,0 +1,25 @@
+ */
+
+tailrec fun solution(d: IntArray, budget: Int): Int = when {
+    d.fold(0, { total, next -> total + next }) <= budget -> d.size
```
여기는 `fold` 대신 그냥 `sum`을 쓰시면 코드를 이해하기 더 쉬울 거에요.

재귀를 쓴다면 그냥 처음에 `sum`을 한번 때리고, `last`로 마지막 값을 구해서 하나씩 빼도 됩니다. `budget`을 관리하기 껄끄럽다면 `gap`이나 뭔가 넘치는 걸로 잡고 줄여갈 수 있죠.

```kotlin
fun solution(d: IntArray, budget: Int) = step(d.sorted(), d.sum() - budget)

fun step(d: List<Int>, gap: Int): Int = when {
    gap <= 0 -> d.size
    else -> step(d.dropLast(1), gap - d.last())
}
```

wholemann committed:
Change the type of d from intarray to collection

ahastudio commented on budget-csh-lyh/src/test/kotlin/com/codingdojo/budget/SolutionTest.kt:
```diff
@@ -0,0 +1,25 @@
+    fun `it returns maximum supportable department count`() {
+        assertThat(solution(mutableListOf(1, 3, 2, 5, 4), 9)).isEqualTo(3)
+        assertThat(solution(mutableListOf(1, 4, 2, 4, 3), 10)).isEqualTo(4)
+        assertThat(solution(mutableListOf(2, 2, 3, 3), 10)).isEqualTo(4)
```
`Collection`으로 받으면 그냥 `arrayOf` 쓰셔도 됩니다.

wholemann replied:
arrayOf는 빨간 줄이 생기네요ㅠ
listOf, mutableListOf, arrayListOf를 해야 되는 듯 합니다.
Kotlin의 권장대로면 listOf가 맞겠죠??? 기본적으로 immutableList를 쓰는게 default니까요.

ahastudio replied:
그럼 `listOf`를 쓰는 게 좋을 거에요.

wholemann replied:
최신 커밋엔 Collection 대신 List<Int>로 타입을 변경하였습니다.

dropLast()는 Collection 타입에선 쓸 수 없어서 이 또한 불필요한 List<Int>

캐스팅이 요구되기 때문에 애초에 List<Int>로 받았습니다.

wholemann committed:
Change the accumulation method from fold to sum

Change the type of d from Collection to List.
The reason is to avoid unnecessary type casting.
